### PR TITLE
HHH-14608 Merge causes StackOverflow when JPA proxy compliance is enabled

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-14608

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/BasicLazyInitializer.java [hunk 1]
```diff
@@ -89,6 +89,19 @@ else if ( method.equals( setIdentifierMethod ) ) {
 
 	}
 
+	Serializable getIdentifier(boolean initialize) {
+		if ( initialize ) {
+			initialize();
+		}
+		return getIdentifier();
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/proxy/pojo/BasicLazyInitializer.java:96]: We might well improve code readability a little bit by changing the method parameter to 'needsToInitialize' or stuff like that, so causal code reader could  grasp the rough meaning solely based on method signature without digging into impl details in method body.

Not sure how to improve the method name of 'isInitializeProxyWhenAccessingIdentifier' though.
Feel free to ignore my comment. Not a big deal.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/proxy/pojo/BasicLazyInitializer.java:96]: Thanks @NathanQingyangXu , I renamed it `needsToInitializeProxy`

**Comment by gbadner** [comment]
IMO, the logic should remain in `AbstractLazyInitializer` so if/when a new `LazyInitializer` extends `AbstractLazyInitializer` (but not BasicLazyInitializer`), it will work as expected with `hibernate.jpa.compliance.proxy=true`.

With the proposed change, `MapLazyInitializer` is no longer affected by `hibernate.jpa.compliance.proxy=true` since it doesn't extend `BasicLazyInitializer`. Obviously, dynamic maps are not defined by JPA, so if an application uses them, they likely would not set `hibernate.jpa.compliance.proxy=true`.

My preference would be to add a method to `AbstractLazyInitializer` that does not have the initialization logic in it for Hibernate's use only; e.g., `#getIdentifierInternal`.

**Comment by dreab8** [comment]
Thanks @gbadner for the suggestion, what do you think about this https://github.com/hibernate/hibernate-orm/pull/3997 ?

**Comment by gbadner** [comment]
Superseded by https://github.com/hibernate/hibernate-orm/pull/3997.

**Pull request closed without merging**